gridstatus/pjm.py

```diff
@@ -538,6 +598,446 @@ def _get_key(self):
+                            (
+                                k1,
+                                k2,
+                            ),
+                        )
+
+        for k1, k2 in pairs:
+            df = self._parse_lmp_series(data[k1][k2])
+            print(f"self._parse_lmp_series(data['{k1}']['{k2}'])")
+            print(df)
+
+    @staticmethod
+    def _dv_lmp_extract_chart_ids(response, verbose=False):
+        html = response.content
+        doc = bs4.BeautifulSoup(html, "html.parser")
+        scripts = doc.find_all("script", {"nonce": True})
+
+        chart_source_id = None
+        menu_form = doc.find("form", {"id": "menuForm"})
+        grandparent = menu_form.parent.parent
+        hidden_elements = grandparent.find_all("input", {"type": "hidden"})
+        j_hidden_elements = [
+            elem for elem in hidden_elements if elem["name"].startswith("j_")
+        ]
+        seed = j_hidden_elements[0]["name"]
+        for script in scripts:
+            lines = script.text.split("\n")
+            for line in lines:
+                match = re.search(rf's:"({seed}[^"]+)"', line)
+                if match is not None:
+                    chart_source_id = match.group(1)
+                    chart_parent_source_id = chart_source_id.split(":")[0]
+
+        if verbose:
+            print(f"chart_source_id = {chart_source_id}", file=sys.stderr)
+            print(f"chart_parent_source_id = {chart_parent_source_id}", file=sys.stderr)
+
+        if chart_source_id is not None and chart_parent_source_id is not None:
+            return {
+                "chart_source_id": chart_source_id,
+                "chart_parent_source_id": chart_parent_source_id,
+            }
+        else:
+            raise ValueError(
+                "Could not get LMP Chart IDs (chart_source_id or chart_parent_source_id)",  # noqa E501
+            )
+
+    @staticmethod
+    def _df_deduplicate(dfs, unique_cols, keep_field, keep_value, verbose=False):
+        """Concatenate dataframes and deduplicate based on a list of columns,
+        keeping keep_field=keep_value.
+        """
+        df = pd.concat(dfs)
+        if verbose:
+            print(f"Starting with {len(df)} rows", file=sys.stderr)
+
+        keep_fields = sorted(df[keep_field].unique().tolist())
+        if keep_value in keep_fields:
+            if keep_fields[0] == keep_value:
+                dedupe_keep = "first"
+            elif keep_fields[-1] == keep_value:
+                dedupe_keep = "last"
+            else:
+                raise ValueError(
+                    f"keep_value {repr(keep_value)} must be "
+                    f"first or last for de-duplication to work: {keep_fields}",
+                )
+
+            # Extract subset without duplicates
+            df.sort_values(keep_field, ascending=True, inplace=True)
+            df.drop_duplicates(subset=unique_cols, keep=dedupe_keep, inplace=True)
+
+        df.reset_index(inplace=True, drop=True)
```

Comment:
this function is only on called in one place and that place does a sort and reset index right after calling it. not sure this reset index is necessary.